### PR TITLE
Add test rc GitHub workflow

### PR DESCRIPTION
**Pull request by Lee-W**
Automate the RC testing process through the GitHub Actions workflow.

## Input
It can accept one of the following two arguments.

* rc_testing_branch: existing rc testing branch (e.g., `rc-test-2023-05-20T05-10-46UTC`)
* issue_url: the URL to the airflow issue that announces rc testing

image

## Workflow steps

image

### CODE REVIEW

**Comment by Lee-W** [comment]
image

**Comment by Lee-W** [comment]
TODO: fix `trigger master dag` job

File: .github/workflows/test-rc-release.yaml [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+on:  # yamllint disable-line rule:truthy
+  workflow_dispatch:
+    inputs:
+      rc_testing_branch:
+        # If a branch is given, the workflow will use it for deployment and testing.
+        # If no branch is provided, the workflow will create a new rc testing branch
+        # for deployment and testing.
+        description: |
+          rc_testing_branch: existing testing branch
+          (One and only one of rc_testing_branch and issue_url is required.)
```
Comment (pankajkoti) [inline .github/workflows/test-rc-release.yaml:13]:
```suggestion
          (Either rc_testing_branch or issue_url is required. You cannot give both.)
```

File: .github/workflows/test-rc-release.yaml [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        required: false
+        default: ''
+      issue_url:
+        description: |
+          issue_url: the URL of the github issue that announce provider testing
```
Comment (pankajkoti) [inline .github/workflows/test-rc-release.yaml:18]:
```suggestion
          issue_url: the GitHub issue URL that tracks the status of Providers release
```

File: .github/workflows/test-rc-release.yaml [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+          (One and only one of rc_testing_branch and issue_url is required.)
```
Comment (pankajkoti) [inline .github/workflows/test-rc-release.yaml:19]:
```suggestion
          (Either rc_testing_branch or issue_url is required. You cannot give both.)
```

File: .github/workflows/test-rc-release.yaml [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        required: false
+      base_git_rev:
+        description: 'The base git revision to test rc release'
```
Comment (pankajkoti) [inline .github/workflows/test-rc-release.yaml:22]:
```suggestion
        description: 'The base git revision to test Providers RCs'
```

**Review by phanikumv — APPROVED** [review]

File: .github/workflows/test-rc-release.yaml [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        required: false
+        type: string
+        default: 'main'
+
+jobs:
+  validate-manual-input:
+    runs-on: 'ubuntu-20.04'
+    steps:
+      - name: Validate user input
+        if: ${{ (inputs.rc_testing_branch != '') == (inputs.issue_url != '') }}
```
Comment (pankajkoti) [inline .github/workflows/test-rc-release.yaml:32]:
```suggestion
        if: ${{ (inputs.rc_testing_branch != '') && (inputs.issue_url != '') }}
```
Should this be an AND? Equal operator seems confusing here.

File: .github/workflows/test-rc-release.yaml [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        run: |
+          echo "One and only one of rc_testing_branch and issue_url is required."
```
Comment (pankajkoti) [inline .github/workflows/test-rc-release.yaml:34]:
```suggestion
          echo "Either rc_testing_branch or issue_url is required. You cannot give both."
```

File: .github/workflows/test-rc-release.yaml [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    needs: wait-for-deployment-to-be-ready
+    if: |
+      always() &&
+      needs.wait-for-deployment-to-be-ready.result == 'success'
+    runs-on: 'ubuntu-20.04'
+    steps:
+      - name: checkout
+        uses: actions/checkout@v3
+        with:
+          ref: ${{ needs.create-branch-for-testing-rc-release.outputs.rc_testing_branch }}
```
Comment (pankajkoti) [inline .github/workflows/test-rc-release.yaml:145]: Shouldn't we use here `export-rc-testing-branch-name` outputs?

**Review by pankajkoti — COMMENTED** [review]
Since, we're taking help of automation, should we also deploy the RCs on the KubernetesExecutor in a follow up PR? Previously, we use to test these only on CeleryExecutor as we did not have the KubernetesExecutor deployment back then. And since we have the KubernetesExecutor recently and it was manual, we quickly only tested it on CE envs. That would save us from discovering late the rare occurrences of providers release failures on it.

File: .github/workflows/test-rc-release.yaml (hunk 5 above)
Comment (Lee-W) [reply .github/workflows/test-rc-release.yaml:32]: This should be `==`. Only one of the value can be true

File: .github/workflows/test-rc-release.yaml (hunk 7 above)
Comment (Lee-W) [reply .github/workflows/test-rc-release.yaml:145]: Yes, fortunately, this won't break our current running workflow🎉. Let me send a PR to fix it now 💪

File: .github/workflows/test-rc-release.yaml (hunk 5 above)
Comment (Lee-W) [reply .github/workflows/test-rc-release.yaml:32]: But if both are not set, it should exit as well

File: .github/workflows/test-rc-release.yaml (hunk 1 above)
Comment (Lee-W) [reply .github/workflows/test-rc-release.yaml:13]: Thanks for your suggestion! This is addressed on https://github.com/astronomer/astronomer-providers/pull/1135

File: .github/workflows/test-rc-release.yaml (hunk 3 above)
Comment (Lee-W) [reply .github/workflows/test-rc-release.yaml:19]: Thanks for your suggestion! This is addressed on https://github.com/astronomer/astronomer-providers/pull/1135

File: .github/workflows/test-rc-release.yaml (hunk 6 above)
Comment (Lee-W) [reply .github/workflows/test-rc-release.yaml:34]: Thanks for your suggestion! This is addressed on https://github.com/astronomer/astronomer-providers/pull/1135

File: .github/workflows/test-rc-release.yaml (hunk 7 above)
Comment (Lee-W) [reply .github/workflows/test-rc-release.yaml:145]: Thanks for your suggestion! This is addressed on https://github.com/astronomer/astronomer-providers/pull/1135

File: .github/workflows/test-rc-release.yaml (hunk 4 above)
Comment (Lee-W) [reply .github/workflows/test-rc-release.yaml:22]: Thanks for your suggestion! This is addressed on https://github.com/astronomer/astronomer-providers/pull/1135

File: .github/workflows/test-rc-release.yaml (hunk 5 above)
Comment (pankajkoti) [reply .github/workflows/test-rc-release.yaml:32]: got it. 
But I think the expression should be explicit.
${{ (inputs. rc_testing_branch == '' && inputs. issue_url == '') || (inputs. rc_testing_branch != '' && inputs. issue_url != '') }}

Comment (Lee-W) [reply .github/workflows/test-rc-release.yaml:32]: Sure. I don't have a strong opinion on it. Just fix it in https://github.com/astronomer/astronomer-providers/pull/1135/commits/d65d8c1a7358f03170ce2ec1043714a2e0ce549b

Comment (pankajkoti) [reply .github/workflows/test-rc-release.yaml:32]: thank you for your consideration and sorry for my strong opinion :)

**Pull request merged**